borderless/explainer.md

```diff
@@ -0,0 +1,512 @@
+- Support for mobileOSs/macOS/windows
+- Support for non-IWAs (Isolated Web Apps)
+
+## Proposal
+
+The solution proposed consists of the following parts:
+
+1. A new `display_override` option `borderless` for the web app manifest.
+2. Enabling draggable regions when in borderless mode.
+3. CSS display-mode media query for borderless.
+4. Leveraging Window Management API permission (_earlier “Multi-screen Windows
+   Placement” API permission_) for borderless mode.
+5. Check that the app is Isolated Web App.
+
+### 1. Setting the `display_override` to borderless
+
+To provide the maximum addressable area for web content, the User Agent (UA)
+will create a frameless window removing all UA provided Chrome, leaving only
+resizing of the window from its borders. The removed window controls will be
+enabled using AWC and HTML/JavaScript/CSS (see non-goals and
+[additional-windowing-controls explainer](<[http://go/additional-windowing-controls](https://github.com/ivansandrk/additional-windowing-controls/blob/main/awc-explainer.md)>)).
+Additionally to hiding the native title bar, also the menu button is hidden.
+
+Example apps in borderless mode could look e.g. like below but the appearance
+would eventually fully depend on what the developer would implement.
+
+![Example borderless PWA](./images/borderless-app-with-AWC.png)
+
+The desire to remove the title bar will be declared within the web app manifest
+by setting the `display_override` to `borderless`. The display_override value
+will be ignored on unsupported OSs and they will follow the normal display mode
+hierarchy.
+
+```
+{
+  "display_override": [ "borderless" ]
+}
+```
+
+### 2. Enabling draggable regions
+
+When the app enters the borderless mode,
+[draggable regions](https://github.com/WICG/window-controls-overlay/blob/main/explainer.md#defining-draggable-regions-in-web-content)
+`-webkit-app-region: drag;` will be enabled.
+
+### 3. CSS display-mode media query for borderless
+
+Similarly to other display modes, borderless should be queryable with @media.
+
+Example media query with borderless:
+
+```
+@media (display-mode: borderless) {
+  .any-css-class-name {
+    margin: 5px;
+  }
+}
+```
+
+See:
+[https://developer.mozilla.org/en-US/docs/Web/CSS/@media/display-mode](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/display-mode)
+
+### 4. Leveraging Window Management permission for borderless
+
+In order for the borderless mode to activate, the app needs to have Window
+Management permission granted. The permission prompt can be triggered using
+JavaScript.
+
+For the **managed** context, permission prompts can be by-passed with a policy
+set by an admin and later on edited in the app’s settings. For the **unmanaged**
+context, the Window management API permission must be granted.
+
+#### Enabling Window Management
+
+Granting the window management API permission can be done with the following
+JavaScript:
+
+```
+const screenDetails = await window.getScreenDetails();
```

Comment:
I have a lot of reservations about how granting the permission (in the unmanaged case) works from an API and UI perspective.

1. There's a weirdness here about having to use JavaScript to ask for permission to use a feature declared in the manifest. This suggests further to me that it shouldn't be tied to the Window Management API permission.
2. What happens if the permission has not yet been granted or denied? (Presumably, we fall back to the next display mode in the chain, e.g. standalone., which means in the VDI case showing a double-title-bar.) Is this acceptable UX?
3. What happens at the instant the permission is granted? Do we expect the window's title bar to disappear upon being given permission? Or does it just change upon the next reload?
4. **My biggest concern**: If the window simply never showed the native title bar, it would be just a fact that this app doesn't have certain controls, and as you say below, there will be alternative ways to get to those features. However, having the app first appear with the native title bar, then a permission prompt being granted causing the title bar to disappear, leads the user to wonder where it went and how to get it back. I don't think it's good to provide UI (in the form of the permission prompt) to make it go away but no way to get it back (other than digging in app settings to revoke the permission). We should address this and possibly provide a way to "reveal" the hidden title bar.

If we're going to show the native title bar to the user on first run anyway, and the user will have to take action to make the native title bar go away, we could essentially make the "permission" a part of the UI instead of something you are prompted for. This is basically the same thought process we went through for WCO: if you're going to start with the full title bar shown, and the user will explicitly give permission to hide it, why not just make that permission a button in the UI, and why not make that button a toggle so they can get it back using the same mechanism they used before.

You could then show an animation when the user toggles the title bar away of the title bar collapsing into the system shelf (indicating where to go to find the controls), or perhaps just collapsing into a small clickable zone in the corner which lets you get it back.

None of this precludes having an admin policy to enable borderless by default. But in the unmanaged case (or managed case where the admin hasn't set that policy), it would be good to think about the UX of how the user grants and un-grants this, and it may make sense for it to not be a normal permission prompt.

Reply:
>> 1. There's a weirdness here about having to use JavaScript to ask for permission to use a feature declared in the manifest. This suggests further to me that it shouldn't be tied to the Window Management API permission.
This could be prompted on window onload.

>> 2. What happens if the permission has not yet been granted or denied? (Presumably, we fall back to the next display mode in the chain, e.g. standalone., which means in the VDI case showing a double-title-bar.) Is this acceptable UX?
Yeah it displays standalone until we have the permission.

>> 3. What happens at the instant the permission is granted? Do we expect the window's title bar to disappear upon being given permission? Or does it just change upon the next reload?
It immeadiately updates the window from standalone to borderless without having to refresh manually. If we toggle the permission from the settings, then it shows the default "Reload" infobar (which is default behaviour for toggling any other permission as well).

>> 4. My biggest concern: If the window simply never showed the native title bar, it would be just a fact that this app doesn't have certain controls............
I think here we have trusted on the first version not being available to other but the trusted VDI partners who we'd trust to not do evil things. This is because IWAs in general are not available in the open web yet. This should be something to take into consideration in the future though if extending it for non-partners. Do you think that should be something to already discuss in the explainer despite it would probably be implemented much later?

Reply:
1. Is not entirely satisfying to me but I understand it would work mechanically. Having a permission prompt that doesn't actually give the app any new API capabilities, it just makes something you declared in your manifest start working, doesn't feel like a permission to me. It feels like a UI setting. (Again, that's how WCO was designed. We could have made you do a permission prompt to activate WCO but we designed it as a UI setting instead.)
2. OK.
3. OK.
4. I don't think that addresses the question. I asked about what if the user grants the permission (which causes the window UI to disappear) and then wants to get it back again. There's no "evil thing" happening here, it's just a natural thing for a user setting to be reversible. The fact that, upon granting the permission, you will be hiding the entire permission granting UI somewhere else, means it can be tricky for a user to try and undo that change.

These things can probably be sorted out after the explainer lands in the repo, so I won't block the review on it. But I do want to keep talking about what this UI looks like in the non-enterprise non-VDI case. We should have a design in mind that scales beyond a set of trusted partners.

Reply:
I already raised this Q up to our PM so we'll keep working on this. I will leave these comments open here.